types/specification.d.ts

```diff
@@ -0,0 +1,105 @@
+/** Stamp Specification v1.6 */
+declare namespace Specification {
+  /** Base type for `properties` */
+  type PropertyMap = { [s: string]: any; };
+
+  /** Base type for `methods` */
+  type MethodMap = { [s: string]: (...args: Array<any>) => any; };
+
+  /** Extract the instance type from Factory, Descriptor (or Instance) */
+  type StampType<I> = I extends Factory<infer I> ? I
+                    : I extends Descriptor<infer I> ? I
+                    : I extends Instance ? I
+                    : never;
+
+  /** A composable object (`stamp` factory or Descriptor) */
+  type Composable<I, U = StampType<I>> = Factory<U> | Descriptor<U>;
+
+  /** Base `stamp` instance */
+  type Instance = { [s: string]: any; };
+
+  /**
+   * A factory function that will produce new objects using the
+   * prototypes that are passed in or composed.
+   */
+  // TODO: interface Factory<I, F extends Factory<any> = Factory<I>>
+  interface Factory<I> {
+      /**
+       * Invokes the `stamp` and returns a new object instance.
+       * @param options Properties you wish to set on the new object.
+       * @param args The remaining arguments are passed to all `.init()` functions.
+       * **WARNING** Avoid using two different `.init()` functions that expect different arguments.
+       * `.init()` functions that take arguments should not be considered safe to compose
+       * with other `.init()` functions that also take arguments. Taking arguments with
+       * an `.init()` function is an anti-pattern that should be avoided, when possible.
+       * @return A new object composed of the Stamps and prototypes provided.
+       */
+      (options?: PropertyMap, ...args: Array<any>): I;
+
+      /** Stamp metadata/composer function */
+      compose: ComposeMethod<I> & Descriptor<I>;
+  }
+
+  /** Create a new stamp based on this descriptor */
+  interface ComposeMethod<I> {
```

Comment:
Can I suggest a better name? This thing is used both - as a standalone function and as a method. Should we rename it to `ComposeFunction`?

Reply:
in Specification it is referred to as the `.compose()`method.
I did not include the "Standalone `compose()` pure function (optional)" in `specification.d.ts`at this point.